### PR TITLE
Adding c_aligned_alloc and supporting functions

### PR DESCRIPTION
**Pull request by mppf**
This PR supersedes #13188 which included discussion of the direction.
  
It adds `c_aligned_alloc` to the CPtr module, which is meant to be 
similar to C17's `aligned_alloc`. It adds `chpl_here_aligned_alloc` for 
locale models to override the behavior. For now these locale models call
`chpl_mem_memalign`.

Future Work:
 * rename `chpl_mem_memalign` `chpl_mem_aligned_alloc`, including in 
   qthreads code

Reviewed by @ronawho with significant input from @gbtitus - thanks!

- [x] full local testing
- [x] new tests pass with quickstart and standard config
- [x] new tests pass on knl

### CODE REVIEW

**Comment by ronawho** [comment]
> Should we rename the runtime memory function chpl_memalign to chpl_aligned_alloc and have the system version call C's aligned_alloc, when it is available?

I'm in favor of renaming the runtime function. aligned_alloc has pretty restrictive alignment requirements (size must be a multiple of alignment) so we'd want to make sure our existing calls adhere to that our we could round in our wrappers.  Let me take a look around the runtime and qthreads for aligned alloc calls. I also need to refresh myself on the alignment requirements for memalign/posixmemalign to make sure we can implement aligned alloc when those are the calls we have available. I probably won't get to until late next week at the earliest and I'd like to chat with Greg about this too.

File: runtime/include/chpl-mem.h [hunk 1]
```diff
@@ -81,6 +81,21 @@ void* chpl_mem_alloc(size_t size, chpl_mem_descInt_t description,
   return chpl_mem_allocMany(1, size, description, lineno, filename);
 }
 
+// assumes alignment is a power of 2
+// assumes that size is a multiple of alignment
+static inline
+void* chpl_mem_aligned_alloc(size_t alignment, size_t size,
```
Comment (ronawho) [inline runtime/include/chpl-mem.h]: We have chpl_mem_memalign. I think we should just call that

Comment (mppf) [reply runtime/include/chpl-mem.h]: I think you're pointing out that this new function is duplicative with chpl_mem_memalign. However if we are renaming chpl_memalign then we would presumably "rename" chpl_mem_memalign to chpl_mem_aligned_alloc, right?

**Comment by ronawho** [comment]
Most of the calls from qthreads do not meet the requirement that the size is a multiple of alignment.   If we want to go with the runtime name change, I think we'll have to round the size up automatically.

**Comment by mppf** [comment]
Rounding up the size of alignment automatically seems reasonable to me...

**Comment by ronawho** [comment]
I'm finding the C/C++ aligned_alloc to be really unhelpful. The size must be a multiple of alignment, which seems overly restrictive. The alignment "Must be a valid alignment supported by the implementation", which is really unhelpful. I don't really think aligned_alloc makes aligned allocation any more portable, but I do like the name better.

I would on board with supporting a chpl_{mem_}aligned_alloc interface that drops the size-is-multiple-of-alignment requirement (and internally we can make that adjustment ourselves.) For the alignment requirement I think we should require a power of 2 that is a multiple of `sizeof(void*)`. This matches posix_memalign, and I'm having a hard time imagining that any other system would have stronger requirements.

**Comment by gbtitus** [comment]
I don't mind adding a Chapel wrapper for `aligned_alloc()` although I'm in agreement with @ronawho that the "size is a multiple of alignment" restriction isn't all that useful.  I definitely don't think we should have something whose name implies it's like C's `memalign` but yet it has the requirements associated with `aligned_alloc` instead.  If both of those are useful then we should just support both `chpl_mem_memalign()` and `chpl_mem_aligned_alloc()`.

**Comment by mppf** [comment]
@gbtitus - do you feel OK about having `chpl_mem_aligned_alloc` which relaxes the "size is a multiple of alignment" ? (In other words this would actually behave more like `memalign` but have a better name. I'm of the opinion that people are unlikely to be upset that our function works in cases when the C function would not).

**Comment by gbtitus** [comment]
> do you feel OK about having `chpl_mem_aligned_alloc` which relaxes the "size is a multiple of alignment" ?

No.  I feel that if we're going to do something that looks like we're wrapping the existing C functions (because our names are theirs with a prefix added) then our wrappers should have the same semantics as those C functions, even if we think those C functions were poorly named or had poorly chosen semantics to begin with.  Obfuscating a flawed interface doesn't make it better.

**Comment by mppf** [comment]
@gbtitus - the goal here is to have a Chapel-user-facing function for aligned allocation. Given the discussion above and the proposal in the PR, how would you suggest we proceed? I don't personally see a sane path that incorporates all of the feedback. Thanks.

**Comment by gbtitus** [comment]
The best solution would be to support `chpl_mem_memalign()`, with the semantics of `memalign()`.  I don't see anyone arguing that Chapel needs the `aligned_alloc()` restriction that size be a multiple of alignment.  I'm sympathetic to the argument that `aligned_alloc` is a better name than `memalign`, but that die is cast and we should just re-use the existing though non-ideal name.  I don't think we should have a function whose name implies that it's like one C interface routine but whose semantics are those of another, though.  So `chpl_mem_aligned_alloc()` with the semantics of `memalign()` is my least favorite.

**Comment by mppf** [comment]
@gbtitus - right but are you just talking about the runtime interface? Or the `CPtr` module too?

My concern with literally adding `c_memalign` to the CPtr module next to `c_malloc` etc is that people who need to use or who are reading code using it will not know it is an allocation routine.

What if we called it, say, `c_malloc_aligned` or `c_malloc_memalign`? These names do not match any existing C function so arguably we can define their behavior. We would then document that it calls a C `memalign` or `posix_memalign` function. (But... I'm not sure it even matters since when using jemalloc or other allocator it calls whatever the implementation needs it to). In that scenario, I would expect that the runtime API would continue to use `chpl_mem_memalign` and we would not try to add `chpl_mem_aligned_alloc`.

**Comment by gbtitus** [comment]
> ... are you just talking about the runtime interface? Or the `CPtr` module too?

Yes, `module CPtr` also.  Since it's an interface to C pointer types and the operations upon them, we should use the existing C names.

**Comment by ronawho** [comment]
>> do you feel OK about having chpl_mem_aligned_alloc which relaxes the "size is a multiple of alignment" ?

> No. I feel that if we're going to do something that looks like we're wrapping the existing C functions (because our names are theirs with a prefix added) then our wrappers should have the same semantics as those C functions, even if we think those C functions were poorly named or had poorly chosen semantics to begin with. Obfuscating a flawed interface doesn't make it better.

Personally, relaxing the requirements seems ok to me. If somebody thinks the behavior is exactly the same as C aligned_alloc, then they will ensure size is a multiple of alignment which won't be a problem.

If somebody tries to use a size that doesn't match we'll fix it for them.

I certainly wouldn't want to have a version with stronger requirements than the C version, but I don't see the problem with reusing a name with less stringent requirements.

**Comment by gbtitus** [comment]
> Personally, relaxing the requirements seems ok to me. If somebody thinks the behavior is exactly the same as C aligned_alloc, then they will ensure size is a multiple of alignment which won't be a problem.
> 
> If somebody tries to use a size that doesn't match we'll fix it for them.

But these require us to second-guess the user's desires or intent.  In the first case, we're saying that users cannot possibly want to allocate aligned things smaller than the alignment, such as for example a regular 4k page at a 1g boundary.  And in the second case, we're deciding that they cannot possibly want the error checking `aligned_alloc()` provides, which would cause a `NULL` return for a too-small size.  I don't think we're in a position to know the user didn't want the behavior they seemed to be asking for.

**Comment by ronawho** [comment]
I'm not sure what a user would do with that error checking beyond going "oh, I forgot to make my size a multiple of alignment". That's not to say there isn't a valid use-case, but I can't think of a use where you'd want to try to do an aligned allocation and fallback to a different allocation scheme when you got the size wrong. 

There is also a proposal to drop the size-is-a-multiple-of-alignment requirement though I don't know what the status of it is -- http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2072.htm

And FWIW I don't have a strong preference for what we do here, but it still doesn't bother me if we end up relaxing restrictions.

**Comment by mppf** [comment]
According to this draft C17 standard
https://web.archive.org/web/20181230041359if_/http://www.open-std.org/jtc1/sc22/wg14/www/abq/c17_updated_proposed_fdis.pdf

the size-is-multiple-of-alignment constraint is removed:

> The aligned_alloc function  allocates  space  for  an  object  whose  alignment  is  specified  by alignment, whose size is specified by size, and whose value is indeterminate.  If the value of alignment is not a valid alignment supported by the implementation the function shall fail by returning a null pointer.

As a result, I think it would be reasonable for us to implement c_aligned_alloc and follow the *C17* definition of that function for the rules about using it.

If we so choose, we can document more specific restrictions on size/alignment that match the Chapel implementation of this function (for our memory layer) - such as @ronawho's suggestion from above which I repeat here:

> For the alignment requirement I think we should require a power of 2 that is a multiple of sizeof(void*). This matches posix_memalign, and I'm having a hard time imagining that any other system would have stronger requirements.

**Comment by gbtitus** [comment]
> ... implement c_aligned_alloc and follow the _C17_ definition of that function ...

Sounds good to me!   👍

**Comment by mppf** [comment]
@ronawho, @gbt - I believe this is ready for final review. At the moment I'm expecting @ronawho to review it though. Thanks!

**Comment by gbtitus** [comment]
> @ronawho, @GBT - I believe this is ready for final review. At the moment I'm expecting @ronawho to review it though. Thanks!

Hey @mppf I believe ampersand-GBT is somebody else.  I'm ampersand-gbtitus.

File: modules/internal/localeModels/knl/LocaleModel.chpl [hunk 2]
```diff
@@ -576,63 +578,108 @@ module LocaleModel {
...
+  pragma "always propagate line file info"
   proc chpl_here_calloc(size:integral, number:int, md:chpl_mem_descInt_t): c_void_ptr {
+    pragma "fn synchronization free"
     pragma "insert line file info"
-      extern proc chpl_mem_calloc(number:size_t, size:size_t, md:chpl_mem_descInt_t) : c_void_ptr;
+    extern proc chpl_mem_calloc(number:size_t, size:size_t, md:chpl_mem_descInt_t) : c_void_ptr;
+
     if allocatingInHbmSublocale() then
       return hbw_calloc(number.safeCast(size_t), size.safeCast(size_t));
+      // TODO: should this halt instead on returning NULL for size > 0?
```
Comment (ronawho) [inline modules/internal/localeModels/knl/LocaleModel.chpl]: I don't follow this TODO

File: runtime/include/chpl-mem.h [hunk 3]
```diff
@@ -119,6 +119,8 @@ void* chpl_mem_realloc(void* memAlloc, size_t size,
   return moreMemAlloc;
 }
 
+// assumes that alignment/boundary is a power of 2
+// size is not necessarily a multiple of alignment
```
Comment (ronawho) [inline runtime/include/chpl-mem.h]: Add multiple of void* requirement to comment?

File: test/memory/allocation/align.chpl [hunk 4]
```diff
@@ -0,0 +1,17 @@
+use CPtr;
```
Comment (ronawho) [inline test/memory/allocation/align.chpl]: I'm fine with leaving these tests here if you want, but if we think it's worthwhile to have more exposure I'd be ok with `test/runtime/configMatters/mem` or something. We don't have that many memory layers anymore, but this is a directory that we're likely to run on different architectures/OSs that might have different aligned allocation implementations.

**Comment by ronawho** [comment]
Left some minor inline feedback, but I'm on board with the overall approach/implementation.

File: modules/internal/localeModels/knl/LocaleModel.chpl (hunk 2 above)
Comment (mppf) [reply modules/internal/localeModels/knl/LocaleModel.chpl]: chpl_mem_calloc calls chpl_mem_allocManyZero which calls chpl_memhook_malloc_post which calls chpl_memhook_check_post if the allocated pointer is NULL which results in a halt at runtime.

Presumably hbw_calloc can return NULL and will not halt if it does.

Comment (gbtitus) [reply modules/internal/localeModels/knl/LocaleModel.chpl]: I wonder if there's a typo and it's really asking if it should halt if size is **less** than 0.

Comment (ronawho) [reply modules/internal/localeModels/knl/LocaleModel.chpl]: I've created https://github.com/Cray/chapel-private/issues/681 for this. Can you either remove the comments, or link to that issue?

Comment (mppf) [reply modules/internal/localeModels/knl/LocaleModel.chpl]: I've removed the comments, thanks for making the issue.

Comment (mppf) [reply modules/internal/localeModels/knl/LocaleModel.chpl]: I've removed the comments, thanks for making the issue.

**Pull request merged**